cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -170,6 +193,99 @@ protected void addPartition(PartitionKey key, String path, boolean explorable, M
   }
 
   @Override
+  public PartitionConsumerResult consumePartitions(PartitionConsumerState partitionConsumerState) {
+    long[] previousInProgress = partitionConsumerState.getVersionsToCheck();
+    long[] inProgress = tx.getInProgress();
+    Long[] noLongerInProgress = setDiff(previousInProgress, inProgress);
+
+    Iterator<Partition> noLongerInProgressPartitions = Iterators.emptyIterator();
+    if (noLongerInProgress.length > 0) {
+      Long firstElement = noLongerInProgress[0];
+      Long lastElement = noLongerInProgress[noLongerInProgress.length - 1];
+      Scanner scanner = partitionsTable.scanByIndex(PartitionedFileSetDefinition.INDEX_COLUMN,
+                                                    Bytes.toBytes(firstElement),
+                                                    Bytes.stopKeyForPrefix(Bytes.toBytes(lastElement)));
+      noLongerInProgressPartitions = new PartitionIterator(scanner, noLongerInProgress);
+    }
+
+    Long startVersion = partitionConsumerState.getStartVersion();
+    long readPointer = tx.getReadPointer();
+    // no read your own writes
+    Scanner scanner = partitionsTable.scanByIndex(PartitionedFileSetDefinition.INDEX_COLUMN,
+                                                  startVersion == null ? null : Bytes.toBytes(startVersion),
+                                                  Bytes.toBytes(readPointer + 1));
+    Iterator<Partition> partitionIterator = Iterators.concat(noLongerInProgressPartitions,
+                                                             new PartitionIterator(scanner, null));
+    return new PartitionConsumerResult(new PartitionConsumerState(tx.getWritePointer(), inProgress), partitionIterator);
```

Comment:
Ok, that would work. I'm saving tx.getReadPointer + 1, then, because we can already process the readPointer in this scan.